File: tests/ruby/src/T011_HeartbeatGenerator.rb
#!/usr/bin/env ruby

require 'libgmsec_ruby'

load 'Test.rb'


class Test_HeartbeatGenerator < Test

    def initialize(args)
        super(args)
    end


    def test()
        test_create_1
        test_create_2
        test_start
        test_stop
        test_change_publish_rate
        test_set_field
    end


    def test_create_1()
        Libgmsec_ruby::Log::info("Test create 1")

        # Nominal test
        hbgen = Libgmsec_ruby::HeartbeatGenerator::create(get_config(), 5)
        check("HeartbeatGenerator should not be running", hbgen.is_running() == false)
        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen)
    end


    def test_create_2()
        Libgmsec_ruby::Log::info("Test create 2")

        emptyFieldList = Libgmsec_ruby::FieldArray.new()
        standardFields = get_standard_fields()

        # Nominal test
        hbgen1 = Libgmsec_ruby::HeartbeatGenerator::create(get_config(), 5, emptyFieldList)
        hbgen2 = Libgmsec_ruby::HeartbeatGenerator::create(get_config(), 5, standardFields)

        check("HeartbeatGenerator should not be running", hbgen1.is_running() == false)
        check("HeartbeatGenerator should not be running", hbgen2.is_running() == false)

        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen1)
        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen2)
    end


    def test_start()
        Libgmsec_ruby::Log::info("Test start()")

        config = get_config()

        hbgen = Libgmsec_ruby::HeartbeatGenerator::create(config, 1, get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        hbgen.start()

        check("HeartbeatGenerator should be running", hbgen.is_running())

        verify_heartbeat_message(config, 1)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        Libgmsec_ruby::TimeUtil::millisleep(2000)

        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen)
    end


    def test_stop()
        Libgmsec_ruby::Log::info("Test stop()")

        config = get_config()

        hbgen = Libgmsec_ruby::HeartbeatGenerator::create(config, 1, get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        hbgen.start()

        check("HeartbeatGenerator should be running", hbgen.is_running())

        hbgen.stop()

        check("HeartbeatGenerator should not be running", hbgen.is_running() == false)

        # Allow time for the HB-gen thread to stop
        Libgmsec_ruby::TimeUtil::millisleep(2000)

        conn = Libgmsec_ruby::Connection::create(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + get_unique_component())

        for i in 0..3 do
            hbmsg = conn.receive(5000)

            if hbmsg != nil
                check("Unexpectedly received a Heartbeat Message", false)
                Libgmsec_ruby::Message::destroy(hbmsg)
            end
        end

        conn.disconnect()

        Libgmsec_ruby::Connection::destroy(conn)
        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen)
    end


    def test_change_publish_rate()
        Libgmsec_ruby::Log::info("Test change_publish_rate()")

        config = get_config()

        hbgen = Libgmsec_ruby::HeartbeatGenerator::create(config, 1, get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        hbgen.start()

        hbgen.change_publish_rate(2)

        Libgmsec_ruby::TimeUtil::millisleep(2000);

        verify_heartbeat_message(config, 2)

        conn = Libgmsec_ruby::Connection::create(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + get_unique_component())

        Libgmsec_ruby::TimeUtil::millisleep(1000);

        hbgen.change_publish_rate(0)

        Libgmsec_ruby::TimeUtil::millisleep(1000);

        lastMessage = false

        for i in 0..2 do
            hbmsg = conn.receive(5000)

            if hbmsg == nil
                check("Failed to receive Heartbeat Message", false)
            elsif hbmsg.get_integer_value("PUB-RATE") != 0
                Libgmsec_ruby::Message::destroy(hbmsg)
            else
                lastMessage = true 
                Libgmsec_ruby::Message::destroy(hbmsg)
                break
            end
        end

        check("Failed to receive message with PUB-RATE of 0", lastMessage)

        conn.disconnect()
        Libgmsec_ruby::Connection::destroy(conn)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        Libgmsec_ruby::TimeUtil::millisleep(2000)

        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen)
    end


    def test_set_field()
        config = Libgmsec_ruby::Config.new( get_config() )     # make a copy

        # Nominal tests
        hbgen = Libgmsec_ruby::HeartbeatGenerator::create(config, 1, get_standard_fields)

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        hbgen.start()

        Libgmsec_ruby::TimeUtil::millisleep(2000)

        Libgmsec_ruby::Log::info("Test set_field() with a Field")
        hbgen.set_field( Libgmsec_ruby::U16Field.new("PUB-RATE", 2) )
        verify_heartbeat_message(config, 2)

        Libgmsec_ruby::Log::info("Test set_field() with a Field Name and long value")
        hbgen.set_field("PUB-RATE", 3)
        verify_heartbeat_message(config, 3)

        Libgmsec_ruby::Log::info("Test set_field() with a Field Name and double value")
        hbgen.set_field("PUB-RATE", 2.0)
        verify_heartbeat_message(config, 2)

        Libgmsec_ruby::Log::info("Test set_field() with a Field Name and string value")
        hbgen.set_field("PUB-RATE", "3")
        verify_heartbeat_message(config, 3)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        Libgmsec_ruby::TimeUtil::millisleep(2000)

        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen)

        # Off-nominal tests
        Libgmsec_ruby::Log::info("Off-nominal cases...")
        config.add_value("gmsec-msg-content-validate", "true")

        hbgen2 = Libgmsec_ruby::HeartbeatGenerator::create(config, 1, get_standard_fields)

        # Add bogus field using a Field
        begin
            hbgen2.set_field( Libgmsec_ruby::U16Field.new("BOGUS-FIELD", 2) )
            check("An expection was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Message Validation Failed"))
        end

        # Add bogus field using a long value
        begin
            hbgen2.set_field("BOGUS-FIELD", 2)
            check("An expection was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Message Validation Failed"))
        end

        # Add bogus field using a double value
        begin
            hbgen2.set_field("BOGUS-FIELD", 2.0)
            check("An expection was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Message Validation Failed"))
        end

        # Add bogus field using a string value
        begin
            hbgen2.set_field("BOGUS-FIELD", "2")
            check("An expection was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Message Validation Failed"))
        end

        Libgmsec_ruby::Log::info("Cleanup...")
        Libgmsec_ruby::HeartbeatGenerator::destroy(hbgen2)
    end


    def verify_heartbeat_message(config, expectedPubRate)
        begin
            conn = Libgmsec_ruby::Connection::create(config)
            conn.connect()
            conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + get_unique_component() + ".+")

            t1 = 0
            t2 = 0

            for i in 0..8 do
                hbmsg = conn.receive(5000)

                # ignore the first few incoming messages (if any)
                if i < 3
                    if hbmsg != nil
                        Libgmsec_ruby::Message::destroy(hbmsg)
                    end
                    next
                end

                if hbmsg != nil
                    if t1 == 0
                        t1 = Libgmsec_ruby::TimeUtil.get_current_time_s()
                    else
                        t2 = Libgmsec_ruby::TimeUtil.get_current_time_s()

                        delta = t2 - t1
                        if delta < expectedPubRate
                            delta = ((t2 - t1) * 10.0 + 0.5) / 10.0
                        end

                        Libgmsec_ruby::Log::info("Expected rate is: " + expectedPubRate.to_s + ", delta is: " + delta.to_s)

                        check("Unexpected publish rate", expectedPubRate == delta.to_i)

                        t1 = t2
                    end

                    check("Unexpected MESSAGE-TYPE", hbmsg.get_string_value("MESSAGE-TYPE") == "MSG")
                    check("Unexpected MESSAGE-SUBTYPE", hbmsg.get_string_value("MESSAGE-SUBTYPE") == "HB")
                    check("Unexpected PUB-RATE", hbmsg.get_integer_value("PUB-RATE") == expectedPubRate)

                    Libgmsec_ruby::Message::destroy(hbmsg)
                else
                    check("Failed to received Heartbeat Message", false)
                end
            end

            conn.disconnect()
            Libgmsec_ruby::Connection::destroy(conn)
        rescue GmsecException => e
            require("verify_heartbeat_message: " + e.message, false)
        end
    end
end


begin
    test = Test_HeartbeatGenerator.new( ARGV )
    test.run
end

File: tests/python3/src/T011_HeartbeatGenerator.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


class Test_HeartbeatGenerator(Test):

    def test(self):
        self.test_constructor_1()
        self.test_constructor_2()
        self.test_start()
        self.test_stop()
        self.test_change_publish_rate()
        self.test_set_field()


    def test_constructor_1(self):
        lp.log_info("Test constructor 1")

        # Nominal test
        hbgen = lp.HeartbeatGenerator(self.get_config(), 5)
        self.check("HeartbeatGenerator should not be running", hbgen.is_running() == False)

        # Off-nominal test(s)
        try:
            config = lp.Config(self.get_config())
            config.add_value("mw-id", "bogus-mw")
            lp.HeartbeatGenerator(config, 5)
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Unable to load" in str(e))


    def test_constructor_2(self):
        lp.log_info("Test constructor 2")

        emptyFieldList = lp.FieldList()
        standardFields = self.get_standard_fields()

        # Nominal test
        hbgen1 = lp.HeartbeatGenerator(self.get_config(), 5, emptyFieldList)
        self.check("HeartbeatGenerator should not be running", hbgen1.is_running() == False)

        hbgen2 = lp.HeartbeatGenerator(self.get_config(), 5, standardFields)
        self.check("HeartbeatGenerator should not be running", hbgen2.is_running() == False)

        # Off-nominal test(s)
        try:
            config = lp.Config(self.get_config())
            config.add_value("mw-id", "bogus-mw")
            lp.HeartbeatGenerator(config, 5, standardFields)
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Unable to load" in str(e))


    def test_start(self):
        lp.log_info("Test start()")

        config = lp.Config(self.get_config())

        hbgen = lp.HeartbeatGenerator(config, 1, self.get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( lp.StringField("COMPONENT", self.get_unique_component()) )

        hbgen.start()

        self.check("HeartbeatGenerator should be running", hbgen.is_running())

        self.verify_heartbeat_message(config, 1)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        lp.TimeUtil.millisleep(2000)

        # Off-nominal tests
        config.add_value("gmsec-msg-content-validate", "true")

        hbgen2 = lp.HeartbeatGenerator(config, 1, self.get_standard_fields())

        # Add bogus field using a Field
        try:
            hbgen2.set_field( lp.U16Field("BOGUS-FIELD", 2) )
            hbgen2.start()
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Message Validation Failed" in str(e))

        # Add bogus field using a long value
        try:
            hbgen2.set_field("BOGUS-FIELD", 2)
            hbgen2.start()
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Message Validation Failed" in str(e))

        # Add bogus field using a double value
        try:
            hbgen2.set_field("BOGUS-FIELD", 2.0)
            hbgen2.start()
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Message Validation Failed" in str(e))

        # Add bogus field using a string value
        try:
            hbgen2.set_field("BOGUS-FIELD", "2")
            hbgen2.start()
            self.check("An exception was expected", False)
        except Exception as e:
            self.check(str(e), "Message Validation Failed" in str(e))


    def test_stop(self):
        lp.log_info("Test stop()")

        config = self.get_config()

        hbgen = lp.HeartbeatGenerator(config, 1, self.get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( lp.StringField("COMPONENT", self.get_unique_component()) )

        hbgen.start()

        self.check("HeartbeatGenerator should be running", hbgen.is_running())

        hbgen.stop()

        self.check("HeartbeatGenerator should not be running", hbgen.is_running() == False)

        # Allow time for the HB-gen thread to stop
        lp.TimeUtil.millisleep(2000)

        conn = lp.Connection(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + self.get_unique_component())

        for i in range(0,3):
            hbmsg = conn.receive(5000)

            if hbmsg != None:
                self.check("Unexpectedly received a Heartbeat Message", False)
                lp.Message.destroy(hbmsg)


    def test_change_publish_rate(self):
        lp.log_info("Test change_publish_rate()")

        config = self.get_config()

        hbgen = lp.HeartbeatGenerator(config, 1, self.get_standard_fields())

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( lp.StringField("COMPONENT", self.get_unique_component()) )

        hbgen.start()

        hbgen.change_publish_rate(2)

        self.verify_heartbeat_message(config, 2)

        conn = lp.Connection(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + self.get_unique_component())

        lp.TimeUtil.millisleep(1000)

        hbgen.change_publish_rate(0)

        lp.TimeUtil.millisleep(1000)

        lastMessage = False

        for i in range(0,3):
            hbmsg = conn.receive(5000)

            if hbmsg == None:
                self.check("Failed to receive Heartbeat Message", False)

            elif hbmsg.get_integer_value("PUB-RATE") != 0:
                lp.Message.destroy(hbmsg)

            else:
                lastMessage = True
                lp.Message.destroy(hbmsg)
                break

        self.check("Failed to receive message with PUB-RATE of 0", lastMessage)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        lp.TimeUtil.millisleep(2000)


    def test_set_field(self):
        config = lp.Config( self.get_config() )     # make a copy

        standardFields = self.get_standard_fields()

        # Nominal tests
        hbgen = lp.HeartbeatGenerator(config, 1, standardFields)

        # To ensure this heartbeat message is unique, we set the COMPONENT field
        hbgen.set_field( lp.StringField("COMPONENT", self.get_unique_component()) )

        hbgen.start()

        lp.TimeUtil.millisleep(2000)

        lp.log_info("Test set_field() with a Field")
        hbgen.set_field( lp.U16Field("PUB-RATE", 2) )
        self.verify_heartbeat_message(config, 2)

        lp.log_info("Test set_field() with a Field Name and long value")
        hbgen.set_field("PUB-RATE", 3)
        self.verify_heartbeat_message(config, 3)

        lp.log_info("Test set_field() with a Field Name and double value")
        hbgen.set_field("PUB-RATE", 2.0)
        self.verify_heartbeat_message(config, 2)

        lp.log_info("Test set_field() with a Field Name and string value")
        hbgen.set_field("PUB-RATE", "3")
        self.verify_heartbeat_message(config, 3)

        hbgen.stop()

        # Allow time for the HB-gen thread to stop
        lp.TimeUtil.millisleep(2000)


    def verify_heartbeat_message(self, config, expectedPubRate):
        conn = lp.Connection(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.HB." + self.get_unique_component() + ".+")

        t1 = 0
        t2 = 0

        for i in range(0,7):
            hbmsg = conn.receive(5000)

            # ignore the first few incoming messages (if any)
            if i < 3:
                if hbmsg != None:
                    lp.Message.destroy(hbmsg)
                continue

            if hbmsg != None:
                if t1 == 0:
                    t1 = lp.TimeUtil.get_current_time_s()
                else:
                    t2 = lp.TimeUtil.get_current_time_s()

                    delta = t2 - t1
                    if delta < expectedPubRate:
                        delta = ((t2 - t1) * 10.0 + 0.5) / 10.0

                    lp.log_info("Expected rate is: " + str(expectedPubRate) + ", delta is: " + str(delta))

                    self.check("Unexpected publish rate", int(expectedPubRate) == int(delta))

                    t1 = t2

                self.check("Unexpected MESSAGE-TYPE", hbmsg.get_string_value("MESSAGE-TYPE") == "MSG")
                self.check("Unexpected MESSAGE-SUBTYPE", hbmsg.get_string_value("MESSAGE-SUBTYPE") == "HB")
                self.check("Unexpected PUB-RATE", hbmsg.get_integer_value("PUB-RATE") == expectedPubRate)

                lp.Message.destroy(hbmsg)
            else:
                self.check("Failed to received Heartbeat Message", False)

        conn.disconnect()


Test.run('Test HeartbeatGenerator', Test_HeartbeatGenerator())
